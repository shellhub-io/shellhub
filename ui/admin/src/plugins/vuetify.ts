// Styles
// import "@mdi/font/css/materialdesignicons.css";
import "vuetify/styles";

// Vuetify
import { createVuetify } from "vuetify";

export default createVuetify({
  theme: {
    defaultTheme: "dark",
  },
});
