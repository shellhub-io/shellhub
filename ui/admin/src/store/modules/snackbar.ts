import { Module } from "vuex";
import { State } from "./../index";

export interface snackbarMessageAndContentType {
  typeMessage: string;
  typeContent: string;
}

export interface SnackbarState {
  snackbarError: boolean;
  snackbarSuccess: boolean;
  snackbarMessageAndContentType: snackbarMessageAndContentType;
  snackbarCopy: boolean;
}

export const snackbar: Module<SnackbarState, State> = {
  namespaced: true,

  state: {
    snackbarError: false,
    snackbarSuccess: false,
    snackbarMessageAndContentType: { typeMessage: "", typeContent: "" },
    snackbarCopy: false,
  },

  getters: {
    snackbarSuccess: (state) => state.snackbarSuccess,
    snackbarError: (state) => state.snackbarError,
    snackbarMessageAndContentType: (state) => state.snackbarMessageAndContentType,
    snackbarCopy: (state) => state.snackbarCopy,
  },

  mutations: {
    setSnackbarSuccessAction: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarSuccess = true;
    },

    setSnackbarNoContent: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarSuccess = true;
    },

    setSnackbarSuccessDefault: (state) => {
      state.snackbarMessageAndContentType = { typeMessage: "default", typeContent: "" };
      state.snackbarSuccess = true;
    },

    unsetSnackbarSuccess: (state) => {
      state.snackbarSuccess = false;
    },

    setSnackbarErrorLoadingOrAction: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    setSnackbarErrorCustom: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    setSnackbarErrorDefault: (state) => {
      state.snackbarMessageAndContentType = { typeMessage: "default", typeContent: "" };
      state.snackbarError = true;
    },

    setSnackbarErrorLicense: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    unsetSnackbarError: (state) => {
      state.snackbarError = false;
    },

    setSnackbarCopy: (state, value) => {
      state.snackbarMessageAndContentType = { typeMessage: "", typeContent: value };
      state.snackbarCopy = true;
    },

    unsetSnackbarCopy: (state) => {
      state.snackbarCopy = false;
    },
  },

  actions: {
    showSnackbarSuccessAction: ({ commit }, value) => {
      const data = { typeMessage: "action", typeContent: value };
      commit("setSnackbarSuccessAction", data);
    },

    showSnackbarSuccessDefault: ({ commit }) => {
      commit("setSnackbarSuccessDefault");
    },

    unsetShowStatusSnackbarSuccess: ({ commit }) => {
      commit("unsetSnackbarSuccess");
    },

    showSnackbarErrorLoading: ({ commit }, value) => {
      const data = { typeMessage: "loading", typeContent: value };
      commit("setSnackbarErrorLoadingOrAction", data);
    },

    showSnackbarNoContent: ({ commit }) => {
      const data = { typeMessage: "no-content", typeContent: "" };
      commit("setSnackbarNoContent", data);
    },

    showSnackbarErrorAction: ({ commit }, value) => {
      const data = { typeMessage: "action", typeContent: value };
      commit("setSnackbarErrorLoadingOrAction", data);
    },

    showSnackbarErrorCustom: ({ commit }, value) => {
      const data = { typeMessage: "custom", typeContent: value };
      commit("setSnackbarErrorCustom", data);
    },

    showSnackbarErrorLicense: ({ commit }, value) => {
      const data = { typeMessage: "licenseRequired", typeContent: value };
      commit("setSnackbarErrorLicense", data);
    },

    showSnackbarErrorDefault: ({ commit }) => {
      commit("setSnackbarErrorDefault");
    },

    unsetShowStatusSnackbarError: ({ commit }) => {
      commit("unsetSnackbarError");
    },

    showSnackbarCopy: ({ commit }, value) => {
      commit("setSnackbarCopy", value);
    },

    unsetShowStatusSnackbarCopy: ({ commit }) => {
      commit("unsetSnackbarCopy");
    },
  },
};
