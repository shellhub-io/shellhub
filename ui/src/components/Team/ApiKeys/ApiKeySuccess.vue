<template>
  <MessageDialog
    v-model="showDialog"
    @close="close"
    title="API Key created successfully!"
    description="Make sure to copy your key now as you will not be able to see it again."
    icon="mdi-check-circle"
    icon-color="success"
    confirm-text="Copy to clipboard"
    cancel-text="Close"
    cancel-data-test="close-btn"
    confirm-data-test="copy-btn"
    @confirm="copyKey"
    @cancel="close"
    data-test="api-key-success-dialog"
  >
    <div class="px-4">
      <v-text-field
        :model-value="apiKey || ''"
        variant="outlined"
        readonly
        density="compact"
        class="monospace-field"
        data-test="generated-key-field"
      >
        <template #append-inner>
          <v-btn
            icon="mdi-content-copy"
            color="primary"
            variant="text"
            size="small"
            @click="copyKey"
            data-test="copy-key-icon-btn"
          />
        </template>
      </v-text-field>
    </div>
  </MessageDialog>
</template>

<script setup lang="ts">
import { } from "vue";
import MessageDialog from "@/components/MessageDialog.vue";
import useSnackbar from "@/helpers/snackbar";

interface Props {
  apiKey?: string;
  keyName?: string;
}

const props = withDefaults(defineProps<Props>(), {
  apiKey: "",
  keyName: "",
});
const snackbar = useSnackbar();
const showDialog = defineModel<boolean>({ required: true });

const copyKey = async () => {
  try {
    await navigator.clipboard.writeText(props.apiKey || "");
    snackbar.showSuccess("API Key copied to clipboard!");
  } catch (err) {
    console.error("Failed to copy: ", err);
    snackbar.showError("Failed to copy API key to clipboard.");
  }
};

const close = () => {
  showDialog.value = false;
};
</script>

<style scoped>
.monospace-field :deep(.v-field__input) {
  font-family: 'Roboto Mono', monospace;
  font-size: 14px;
}
</style>
