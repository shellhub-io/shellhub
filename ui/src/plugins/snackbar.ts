import { App, Plugin, reactive } from "vue";

const SnackbarInjectionKey = Symbol("snackbar");

type SnackbarType = "success" | "error" | "info" | "warning";

interface SnackbarState {
  message: string;
  type: SnackbarType;
  show: boolean;
}

interface ISnackbarPlugin {
  showSuccess: (msg: string) => void;
  showError: (msg: string) => void;
  showInfo: (msg: string) => void;
  showWarning: (msg: string) => void;
  getMessage: () => string;
  getType: () => SnackbarType;
  getShow: () => boolean;
}

const state = reactive<SnackbarState>({
  message: "",
  type: "info",
  show: false,
});

const showSnackbar = (type: SnackbarType, message: string) => {
  state.message = message;
  state.type = type;
  state.show = true;

  setTimeout(() => {
    state.show = false;
  }, 4000);
};

const plugin: ISnackbarPlugin = {
  showSuccess: (msg: string) => showSnackbar("success", msg),
  showError: (msg: string) => showSnackbar("error", msg),
  showInfo: (msg: string) => showSnackbar("info", msg),
  showWarning: (msg: string) => showSnackbar("warning", msg),
  getMessage: () => state.message,
  getType: () => state.type,
  getShow: () => state.show,
};

const SnackbarPlugin: Plugin = {
  install(app: App) {
    app.provide(SnackbarInjectionKey, plugin);
  },
};

export { SnackbarPlugin, SnackbarInjectionKey, plugin };
export type { SnackbarState, ISnackbarPlugin };
