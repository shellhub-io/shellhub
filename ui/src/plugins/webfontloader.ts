// webfontloader documentation: https://github.com/typekit/webfontloader
const loadFonts = async () => {
  const webFontLoader = await import("webfontloader");

  webFontLoader.load({
    google: {
      families: ["Roboto:100,300,400,500,700,900&display=swap"],
    },
  });
};

export default loadFonts;
