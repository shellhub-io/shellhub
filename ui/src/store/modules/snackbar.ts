import { Module } from "vuex";
import { State } from "..";

export interface snackbarMessageAndContentType {
  typeMessage: string;
  typeContent: string;
}

export interface SnackbarState {
  message: string,
  type: string,
  snackbarError: boolean;
  snackbarSuccess: boolean;
  snackbarMessageAndContentType: snackbarMessageAndContentType;
  snackbarCopy: boolean;
}

export const snackbar: Module<SnackbarState, State> = {
  namespaced: true,
  state: {
    message: "",
    type: "",
    snackbarError: false,
    snackbarSuccess: false,
    snackbarMessageAndContentType: {
      typeMessage: "",
      typeContent: "",
    },
    snackbarCopy: false,
  },

  getters: {
    snackbarSuccess: (state) => state.snackbarSuccess,
    snackbarError: (state) => state.snackbarError,
    snackbarMessageAndContentType: (state) => state.snackbarMessageAndContentType,
    snackbarCopy: (state) => state.snackbarCopy,
  },

  mutations: {
    showMessage: (state, data) => {
      state.type = data.type;
      state.message = data.message;
    },

    setSnackbarSuccessAction: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarSuccess = true;
    },

    unsetSnackbarSuccess: (state) => {
      state.snackbarSuccess = false;
    },

    setSnackbarSuccessNotRequest: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarSuccess = true;
    },

    setSnackbarErrorLoadingOrAction: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    setSnackbarErrorDefault: (state) => {
      state.snackbarMessageAndContentType = {
        typeMessage: "default",
        typeContent: "",
      };
      state.snackbarError = true;
    },

    unsetSnackbarError: (state) => {
      state.snackbarError = false;
    },

    setSnackbarCopy: (state, value) => {
      state.snackbarMessageAndContentType = {
        typeMessage: "",
        typeContent: value,
      };
      state.snackbarCopy = true;
    },

    unsetSnackbarCopy: (state) => {
      state.snackbarCopy = false;
    },
  },

  actions: {
    showSnackbarSuccessAction: ({ commit }, value) => {
      const data = {
        typeMessage: "action",
        typeContent: value,
      };
      commit("setSnackbarSuccessAction", data);
    },

    unsetShowStatusSnackbarSuccess: ({ commit }) => {
      commit("unsetSnackbarSuccess");
    },

    showSnackbarSuccessNotRequest: (context, value) => {
      const data = {
        typeMessage: "notRequest",
        typeContent: value,
      };
      context.commit("setSnackbarSuccessNotRequest", data);
    },

    showSnackbarErrorLoading: ({ commit }, value) => {
      const data = {
        typeMessage: "loading",
        typeContent: value,
      };
      commit("setSnackbarErrorLoadingOrAction", data);
    },

    showSnackbarErrorAction: ({ commit }, value) => {
      const data = {
        typeMessage: "action",
        typeContent: value,
      };
      commit("setSnackbarErrorLoadingOrAction", data);
    },

    showSnackbarErrorDefault: ({ commit }) => {
      commit("setSnackbarErrorDefault");
    },

    unsetShowStatusSnackbarError: ({ commit }) => {
      commit("unsetSnackbarError");
    },

    showSnackbarCopy: ({ commit }, value) => {
      commit("setSnackbarCopy", value);
    },

    unsetShowStatusSnackbarCopy: ({ commit }) => {
      commit("unsetSnackbarCopy");
    },
  },
};
