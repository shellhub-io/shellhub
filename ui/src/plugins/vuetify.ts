// Styles
import "@mdi/font/css/materialdesignicons.css";
import "@fortawesome/fontawesome-free/css/all.css";
import "vuetify/styles";
import { VFileUpload, VFileUploadItem } from "vuetify/labs/VFileUpload";
import "../styles/_variables.scss";
import { aliases as faAliases, fa } from "vuetify/iconsets/fa";

// Vuetify
import { createVuetify } from "vuetify";

const light = {
  dark: false,
  colors: {
    primary: "#667acc",
    secondary: "#FFFFFF",
    background: "#F5F5F5",
    tabs: "#F5F5F5",
    foreground: "#F5F5F5",
    paymentForm: "#F5F5F5",
    terminal: "#0f1526",
    "v-theme-background": "#FFFFFF",
    "v-theme-surface": "#FFFFFF",
    "v-theme-card": "#FFFFFF",
  },
};

const dark = {
  dark: true,
  colors: {
    primary: "#667acc",
    secondary: "#1E2127",
    background: "#18191B",
    tabs: "#1E1E1E",
    foreground: "#1E1E1E",
    paymentForm: "#E0E0E0",
    terminal: "#0f1526",
    "v-theme-background": "#1E2127",
    "v-theme-surface": "#1E2127",
    "v-theme-card": "#22252B",
  },
};

export default createVuetify({
  components: {
    VFileUpload,
    VFileUploadItem,
  },
  theme: {
    defaultTheme: "dark",
    themes: {
      dark,
      light,
    },
  },
  defaults: {
    global: {

    },
    VTextField: {
      variant: "outlined",
      density: "comfortable",
    },
    VSelect: {
      variant: "outlined",
      density: "comfortable",
    },
    VTextarea: {
      variant: "outlined",
    },
  },
  icons: {
    faAliases,
    sets: {
      fa,
    },
  },
});
