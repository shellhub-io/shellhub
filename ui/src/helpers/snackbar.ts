import { inject } from "vue";
import { InjectionKey, type ISnackbarPlugin } from "@/plugins/snackbar";

const useSnackbar = (): ISnackbarPlugin => inject(InjectionKey) as ISnackbarPlugin;

export default useSnackbar;
