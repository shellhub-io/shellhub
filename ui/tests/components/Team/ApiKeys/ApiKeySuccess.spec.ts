import { createPinia, setActivePinia } from "pinia";
import { DOMWrapper, flushPromises, mount, VueWrapper } from "@vue/test-utils";
import { createVuetify } from "vuetify";
import { expect, describe, it, beforeEach, vi } from "vitest";
import ApiKeySuccess from "@/components/Team/ApiKeys/ApiKeySuccess.vue";
import { SnackbarPlugin } from "@/plugins/snackbar";

const mockSnackbar = {
  showSuccess: vi.fn(),
  showError: vi.fn(),
  showWarning: vi.fn(),
  showInfo: vi.fn(),
};

vi.mock("@/helpers/snackbar", () => ({
  default: () => mockSnackbar,
}));

// Mock navigator.clipboard
Object.assign(navigator, {
  clipboard: {
    writeText: vi.fn(),
  },
});

describe("Api Key Success", () => {
  let wrapper: VueWrapper<InstanceType<typeof ApiKeySuccess>>;
  const vuetify = createVuetify();
  setActivePinia(createPinia());

  const defaultProps = {
    modelValue: true,
    apiKey: "test-api-key-12345",
    keyName: "test-key",
  };

  beforeEach(async () => {
    vi.clearAllMocks();
    wrapper = mount(ApiKeySuccess, {
      global: {
        plugins: [vuetify, SnackbarPlugin],
      },
      props: defaultProps,
    });
  });

  it("Is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("Renders the component", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Renders components correctly", async () => {
    const dialog = new DOMWrapper(document.body);
    await flushPromises();

    expect(dialog.find('[data-test="api-key-success-dialog"]').exists()).toBe(true);
    expect(dialog.find('[data-test="generated-key-field"]').exists()).toBe(true);
    expect(dialog.find('[data-test="copy-key-icon-btn"]').exists()).toBe(true);
    expect(dialog.find('[data-test="copy-btn"]').exists()).toBe(true);
    expect(dialog.find('[data-test="close-btn"]').exists()).toBe(true);
  });

  it("Displays the API key in the text field", async () => {
    const dialog = new DOMWrapper(document.body);
    await flushPromises();

    const keyField = dialog.find('[data-test="generated-key-field"] input');
    expect(keyField.element.value).toBe("test-api-key-12345");
  });

  it("Copies API key to clipboard when copy button is clicked", async () => {
    const writeTextSpy = vi.spyOn(navigator.clipboard, "writeText").mockResolvedValue();

    const dialog = new DOMWrapper(document.body);
    await flushPromises();

    const copyButton = dialog.find('[data-test="copy-btn"]');
    await copyButton.trigger("click");
    await flushPromises();

    expect(writeTextSpy).toHaveBeenCalledWith("test-api-key-12345");
    expect(mockSnackbar.showSuccess).toHaveBeenCalledWith("API Key copied to clipboard!");
  });

  it("Copies API key to clipboard when icon button is clicked", async () => {
    const writeTextSpy = vi.spyOn(navigator.clipboard, "writeText").mockResolvedValue();

    const dialog = new DOMWrapper(document.body);
    await flushPromises();

    const iconButton = dialog.find('[data-test="copy-key-icon-btn"]');
    await iconButton.trigger("click");
    await flushPromises();

    expect(writeTextSpy).toHaveBeenCalledWith("test-api-key-12345");
    expect(mockSnackbar.showSuccess).toHaveBeenCalledWith("API Key copied to clipboard!");
  });

  it("Shows error message when clipboard copy fails", async () => {
    const writeTextSpy = vi.spyOn(navigator.clipboard, "writeText").mockRejectedValue(new Error("Clipboard error"));
    const consoleSpy = vi.spyOn(console, "error").mockImplementation(() => {
      // Mock implementation to silence console.error in tests
    });

    const dialog = new DOMWrapper(document.body);
    await flushPromises();

    const copyButton = dialog.find('[data-test="copy-btn"]');
    await copyButton.trigger("click");
    await flushPromises();

    expect(writeTextSpy).toHaveBeenCalledWith("test-api-key-12345");
    expect(mockSnackbar.showError).toHaveBeenCalledWith("Failed to copy API key to clipboard.");
    expect(consoleSpy).toHaveBeenCalledWith("Failed to copy: ", expect.any(Error));

    consoleSpy.mockRestore();
  });

  it("Has close button that triggers close method", async () => {
    const closeSpy = vi.spyOn(wrapper.vm, "close");

    const dialog = new DOMWrapper(document.body);
    await flushPromises();

    const closeButton = dialog.find('[data-test="close-btn"]');
    expect(closeButton.exists()).toBe(true);

    // Test that the close method exists and can be called
    wrapper.vm.close();
    expect(closeSpy).toHaveBeenCalled();
  });

  it("Closes dialog when close method is called directly", async () => {
    expect(wrapper.vm.showDialog).toBe(true);

    wrapper.vm.close();
    await flushPromises();

    expect(wrapper.vm.showDialog).toBe(false);
  });
});
