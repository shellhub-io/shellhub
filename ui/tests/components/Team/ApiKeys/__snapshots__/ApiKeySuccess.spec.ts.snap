// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Api Key Success > Renders the component 1`] = `
"<!--teleport start-->
<!--teleport end-->"
`;
